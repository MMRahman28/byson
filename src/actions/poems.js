import { v4 as uuid } from 'uuid';
import database from '../firebase/firebase';
// ADD_POEM
export const addPoem = (poem) => (
{
    type: 'ADD_POEM',
    poem

  });

export const startAddPoem = (poemData = {}) => {
  return (dispatch) => {
    const {
      note = '',
      createdAt = 0
    } = poemData;

    const poem = {note, createdAt};
  return   database.ref('poems').push(poem).then((ref) => {
      dispatch(addPoem({
        id: ref.key,
        ...poem
      }));
    });
  };
};

// REMOVE_POEM

export const removePoem = ({id} = {}) => (
  {
      type: 'REMOVE_POEM',
      id
  }

);

// EDIT_POEM
export const editPoem = (id, updates) => (
    {
        type: 'EDIT_POEM',
        id,
        updates

    }
);

// SET_POEM
export const setPoem = (poems) => ({
  type: 'SET_POEM',
  poems
});
