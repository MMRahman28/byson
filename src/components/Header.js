import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Poems</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>PoemsPage</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create poem</NavLink>

        <NavLink to="/help" activeClassName="is-active">Help</NavLink>

    </header>
);
export default Header;
