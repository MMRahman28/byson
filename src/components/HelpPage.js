import React from 'react';
const HelpPage = () => (

    <div>
    This is my help expense page.
    </div>
);
export default HelpPage;
