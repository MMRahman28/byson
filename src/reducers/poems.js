import uuid from 'uuid';
//import { v4 as uuid } from 'uuid';
const poemsReducerDefaultState = [];
const randomPoems = [...Array(4)].map(() => Math.floor(Math.random() * 9));
export default (state = poemsReducerDefaultState, action) => {
    switch (action.type)
    {
        case 'ADD_POEM':
            return [...state, action.poem];

        case 'REMOVE_POEM':
            return [...state].filter((poem) => (poem.id !== action.id));
        case 'EDIT_POEM':
            return state.map((poem) => {
                if(poem.id === action.id){

                     return {
                         ...poem,
                         ...action.updates
                     }
                } else {
                    return poem;
                }
            });

        case 'SET_POEMS':
          return action.poems;
          
        // case 'SET_POEM':
        //      return [...state].filter((poem) => (randomPoems.includes(poem.id)));


        default:
            return state;
    }
};
