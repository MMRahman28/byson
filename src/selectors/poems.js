import moment from 'moment';

// expenses: state.expenses
// {text, sortBy, startDate, endate} : state.filters
export default (poems, { note, sortBy, startDate, endDate}) => {
    return poems.filter((poem) => {
         const createdAtMoment = moment(poem.createdAt);
         const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'): true;


        // figure out if expenses.description has the text variable string inside of it
        // includes
        // convert both strings to lowercase
        const textMatch = typeof note !== 'string' || poem.note.toLowerCase().includes(note.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;

    }).sort((a, b) => {
    //   // sortBy -> date
    //   // put the ones with the latest date first
    if(sortBy === 'date') {
             return a.createdAt < b.createdAt ? 1 : -1;
           }
        });
    //     // sortBy -> amount
    //     // put the ones with a greater amount first
    //     else if(sortBy === 'amount') {
    //         return a.amount < b.amount ? 1 : -1;
    //     }
    // });

};
