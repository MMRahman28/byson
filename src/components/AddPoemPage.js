import React from 'react';
import PoemForm from './PoemForm';
import {connect } from 'react-redux';
import { startAddPoem } from '../actions/poems';


export class AddPoemPage extends React.Component {
  onSubmit = (poem) => {
    this.props.startAddPoem(poem);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
          <h1>Add Poem</h1>
          <PoemForm
           onSubmit={this.onSubmit}
          />
      </div>
    );
  }
}

// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//          onSubmit={(expense) => {
//             //props.dispatch(addExpense(expense));
//             props.onSubmit(expense);
//             props.history.push('/');
//          }}
//         />
//     </div>
// );

const mapDispatchToProps = (dispatch) => ({
  startAddPoem: (poem) => dispatch(startAddPoem(poem))
});

export default connect(null, mapDispatchToProps)(AddPoemPage);
