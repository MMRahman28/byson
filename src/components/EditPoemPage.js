import React from 'react';
import { connect } from 'react-redux';
import PoemForm from './PoemForm';
import { editPoem, startRemovePoem } from '../actions/poems';

// Refactor EditExpensePage to be a class based Component
// Setup mapDispatchToProps editExpense and removeExpense




export class EditPoemPage extends React.Component {
  onSubmit = (poem) => {
    // Dispatch the action to edit the expense
    // Redirect to the dashboard
    this.props.editPoem(this.props.poem.id, poem);
    this.props.history.push('/');
  };


  onRemove = () => {
    this.props.startRemovePoem({id: this.props.poem.id});
    this.props.history.push('/');
  };
  render(){
    return (
        <div>
            <PoemForm
             poem={this.props.poem}
             onSubmit={this.onSubmit}
             />
             <button onClick={this.onRemove}>Remove</button>
        </div>

    );
  }
};
const mapStateToProps = (state, props) => ({
  poem: state.poems.find((poem) => poem.id === props.match.params.id)
  //dynamic part of the url (id) will be used as id or vice versa
});

const mapDispatchToProps = (dispatch, props) => ({
  editPoem: (id, poem) => dispatch(editPoem(id, poem)),
  startRemovePoem: (data) => dispatch(startRemovePoem(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPoemPage);
