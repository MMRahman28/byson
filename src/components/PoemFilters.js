import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class PoemFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

 onTextChange = (e) => {
   this.props.setTextFilter(e.target.value);
 };
 onSortChange =(e) => {
   if (e.target.value === 'date') {
     this.props.sortByDate();
   }

 };


  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.note}
          onChange={this.onTextChange}/>
          <select
            value={this.props.filters.sortBy}
            onChange={this.onSortChange}
          >
          <option value="date">Date</option>
          //<option value="Text">Text</option>
          </select>
          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId="SDP"
            endDate={this.props.filters.endDate}
            endDateId="EDP"
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat={() => "DD/MM/YYYY"}
          />

        </div>
    );
  }
};

const mapStateToProps = (state) => ({
    filters: state.filters
  });

const mapDispatchToProps = (dispatch) => ({
   setTextFilter: (note) => dispatch(setTextFilter(note)),
   sortByDate: () => dispatch(sortByDate()),
   setStartDate: (startDate) => dispatch(setStartDate(startDate)),
   setEndDate: (endDate) => dispatch(setEndDate(endDate))
   });

export default connect(mapStateToProps, mapDispatchToProps)(PoemFilters);
