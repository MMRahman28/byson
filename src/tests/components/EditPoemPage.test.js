import React from 'react';
import { shallow } from 'enzyme';
import poems from '../fixtures/poems';
import { EditPoemPage } from '../../components/EditPoemPage';

let editPoem, removePoem, history, wrapper;

beforeEach(() => {
  editPoem = jest.fn();
  removePoem = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(

    <EditPoemPage
      editPoem={editPoem}
      removePoem={removePoem}
      history={history}
      poem={poems[2]}
      />


  );
});

test('should render EditPoemPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editPoem', () => {

  wrapper.find('PoemForm').prop('onSubmit')(poems[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editPoem).toHaveBeenLastCalledWith(poems[2].id, poems[2]);
});

test('should handle removePoem', () => {

  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removePoem).toHaveBeenLastCalledWith({
    id: poems[2].id
  });
});
