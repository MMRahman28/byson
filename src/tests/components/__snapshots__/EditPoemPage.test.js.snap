// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPoemPage correctly 1`] = `
<div>
  <PoemForm
    onSubmit={[Function]}
    poem={
      Object {
        "createdAt": 345600000,
        "id": "3",
        "note": "Evening Poem",
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
